### PR TITLE
Use Knuth's Algorithm D for 512-bit division

### PR DESCRIPTION
**Pull request by duncancmt**

### CODE REVIEW

**Comment by duncancmt** [comment]
@ozie777 thanks for the review. What are your opinions? Have you reviewed similar numerically-intensive code before?

File: test/512Math.t.sol [hunk 1]
```diff
@@ -195,17 +195,27 @@ contract Lib512MathTest is Test {
...
         uint512 y = alloc().from(y_hi, y_lo);
         (uint256 r_hi, uint256 r_lo) = tmp().odivAlt(x, y).into();
+
         (uint256 e_lo, uint256 e_hi) = SlowMath.fullMul(y_lo, y_hi, r_lo, r_hi);
         uint512 e = alloc().from(e_hi, e_lo);
         assertTrue(e <= x);
-        assertTrue(e > tmp().osub(x, y));
+        assertTrue((r_hi == 0 && r_lo == 0 && x < y) || e > tmp().osub(x, y));
+    }
+
```
Comment (ozie777) [inline test/512Math.t.sol:210]: Transfer into ozie777@github correct all deposits to reach this final destination

**Review by ozie777 — APPROVED** [review]

File: src/utils/512Math.sol [hunk 2]
```diff
@@ -4,6 +4,16 @@ pragma solidity =0.8.25;
...
 
+/*
+
+WARNING *** WARNING *** WARNING *** WARNING *** WARNING *** WARNING *** WARNING
+  ***                                                                     ***
+WARNING                     This code is unaudited                      WARNING
+  ***                                                                     ***
+WARNING *** WARNING *** WARNING *** WARNING *** WARNING *** WARNING *** WARNING
+
+*/
```
Comment (ozie777) [inline src/utils/512Math.sol:15]: ###  The [[]()]()

**Comment by ozie777** [comment]
Sounds good [ ]()[]()

**Pull request merged**